### PR TITLE
SerializableItem: implement JSON from string

### PR DESCRIPTION
**Pull request by tigercosmos**
Support JSON from string
- add `from_json` into `MM_DECL_SERIALIZABLE`
- implemented `JsonNode`, which is the token of JSON
- implemented JSON parser: `parse_array`, `parse_object`
- added a helper class: `JsonHelper`

### CODE REVIEW

File: cpp/modmesh/base.hpp
Comment (tigercosmos) [inline cpp/modmesh/base.hpp]: some common helper functions

File: cpp/modmesh/serialization/SerializableItem.cpp
Comment (tigercosmos) [inline cpp/modmesh/serialization/SerializableItem.cpp:42]: print readable error message for parsing json

File: cpp/modmesh/serialization/SerializableItem.cpp
Comment (tigercosmos) [inline cpp/modmesh/serialization/SerializableItem.cpp:97]: for the error message

File: cpp/modmesh/serialization/SerializableItem.cpp
Comment (tigercosmos) [inline cpp/modmesh/serialization/SerializableItem.cpp:110]: a state machine parsing JSON array, same for the `parse_object`

File: cpp/modmesh/serialization/SerializableItem.cpp
Comment (tigercosmos) [inline cpp/modmesh/serialization/SerializableItem.cpp:180]: parse the child, which is a JSON object here.

File: cpp/modmesh/serialization/SerializableItem.cpp
Comment (tigercosmos) [inline cpp/modmesh/serialization/SerializableItem.cpp]: the child is array.

File: cpp/modmesh/serialization/SerializableItem.cpp
Comment (tigercosmos) [inline cpp/modmesh/serialization/SerializableItem.cpp:246]: the basic types go to here.

File: cpp/modmesh/serialization/SerializableItem.cpp
Comment (tigercosmos) [inline cpp/modmesh/serialization/SerializableItem.cpp:185]: not support array in array for now.

File: cpp/modmesh/serialization/SerializableItem.cpp
Comment (tigercosmos) [inline cpp/modmesh/serialization/SerializableItem.cpp:288]: state machine for JSON object.

File: cpp/modmesh/serialization/SerializableItem.cpp
Comment (tigercosmos) [inline cpp/modmesh/serialization/SerializableItem.cpp:395]: nested object.

File: cpp/modmesh/serialization/SerializableItem.cpp
Comment (tigercosmos) [inline cpp/modmesh/serialization/SerializableItem.cpp:428]: the child is array.

File: cpp/modmesh/serialization/SerializableItem.hpp
Comment (tigercosmos) [inline cpp/modmesh/serialization/SerializableItem.hpp:74]: support common types for now.

File: cpp/modmesh/serialization/SerializableItem.hpp
Comment (tigercosmos) [inline cpp/modmesh/serialization/SerializableItem.hpp]: move this function into the helper class.

File: cpp/modmesh/serialization/SerializableItem.hpp
Comment (tigercosmos) [inline cpp/modmesh/serialization/SerializableItem.hpp]: save the value into the class.

File: cpp/modmesh/serialization/SerializableItem.hpp
Comment (tigercosmos) [inline cpp/modmesh/serialization/SerializableItem.hpp:302]: parse the JSON first.

File: cpp/modmesh/serialization/SerializableItem.hpp
Comment (tigercosmos) [inline cpp/modmesh/serialization/SerializableItem.hpp:311]: then we can save the values from the parsed nodes into class's values.

File: cpp/modmesh/base.hpp
Comment (yungyuc) [inline cpp/modmesh/base.hpp]: Use STL `is_integral`: https://en.cppreference.com/w/cpp/types/is_integral

File: cpp/modmesh/base.hpp
Comment (yungyuc) [inline cpp/modmesh/base.hpp]: `is_floating_point`: https://en.cppreference.com/w/cpp/types/is_floating_point

File: cpp/modmesh/base.hpp
Comment (yungyuc) [inline cpp/modmesh/base.hpp]: `isdigit`: https://en.cppreference.com/w/cpp/string/byte/isdigit

File: cpp/modmesh/base.hpp
Comment (yungyuc) [reply cpp/modmesh/base.hpp]: `isalpha`: https://en.cppreference.com/w/cpp/string/byte/isalpha

File: cpp/modmesh/serialization/SerializableItem.cpp
Comment (yungyuc) [inline cpp/modmesh/serialization/SerializableItem.cpp]: Thanks for adding the blank line.

File: cpp/modmesh/serialization/SerializableItem.cpp
Comment (yungyuc) [reply cpp/modmesh/serialization/SerializableItem.cpp:185]: It'll be good to append "yet" in the code comment.  Nested arrays will be needed.

I agree we do not need it badly at the moment.

File: cpp/modmesh/serialization/SerializableItem.cpp
Comment (yungyuc) [reply cpp/modmesh/serialization/SerializableItem.cpp:110]: Sweet. I like it. It's the kind of code I cannot quickly write.

File: cpp/modmesh/serialization/SerializableItem.cpp
Comment (yungyuc) [inline cpp/modmesh/serialization/SerializableItem.cpp:323]: I like the case ending marks.

File: cpp/modmesh/serialization/SerializableItem.hpp
Comment (yungyuc) [inline cpp/modmesh/serialization/SerializableItem.hpp]: Do not add type aliases to smart pointers (`JsonNodePtr`).  Use the modern style `using` (for style consistency):

```cpp
using std::unordered_map<std::string, std::unique_ptr<JsonNode>> JsonMap;
using JsonArray = std::vector<std::unique_ptr<JsonNode>>;
```

Comment (yungyuc) [reply cpp/modmesh/serialization/SerializableItem.hpp]: Aliases to smart pointer types always confuse people.  We will need to look up the definition.  It's better to always spell out the smart pointer type.

File: cpp/modmesh/serialization/SerializableItem.hpp
Comment (yungyuc) [inline cpp/modmesh/serialization/SerializableItem.hpp:106]: The function body of `parse` is a little bit too long for declaration, but it's fine for now.

File: cpp/modmesh/serialization/SerializableItem.hpp
Comment (yungyuc) [inline cpp/modmesh/serialization/SerializableItem.hpp]: I am not familiar with `constexpr if` enough.  Do you need the keyword here?

File: cpp/modmesh/serialization/SerializableItem.hpp
Comment (yungyuc) [reply cpp/modmesh/serialization/SerializableItem.hpp]: I don't like the output argument.  But it looks like necessary evil here.  Is it so?  If yes, is it possible to make it private?

Comment (yungyuc) [reply cpp/modmesh/serialization/SerializableItem.hpp]: And this function `from_json_string` is certainly too long for declaration.  Please make it outside the class declaration.

File: cpp/modmesh/serialization/SerializableItem.hpp
Comment (yungyuc) [inline cpp/modmesh/serialization/SerializableItem.hpp]: This overload can be outside the declaration too.

File: gtests/test_nopython_serializable.cpp
Comment (yungyuc) [inline gtests/test_nopython_serializable.cpp:170]: We will need Python tests for these things.  It's not productive to test for strings in C++.

It can use a new PR or even a new issue.

**Review by yungyuc — CHANGES_REQUESTED** [review]
It's an exciting PR.  Things to fix:

- [x] Use STL `isalpha`, `isdigit`, `is_integral`, and `is_floating_point`.
- [x] Do not alias smart pointer types.
- [x] Use the modern `using` syntax.
- [x] Clarify how to improve the output argument of `from_json_string()`.
- [x] `from_json_string()` is too long for class declaration.

File: cpp/modmesh/serialization/SerializableItem.hpp
Comment (tigercosmos) [reply cpp/modmesh/serialization/SerializableItem.hpp]: `constexpr if` is doing the condition in compilation time, which will not have a runtime branch.

File: cpp/modmesh/serialization/SerializableItem.hpp
Comment (tigercosmos) [reply cpp/modmesh/serialization/SerializableItem.hpp]: > And this function from_json_string is certainly too long for declaration. Please make it outside the class declaration.

I put it in the class because L228 (`value.from_json(std::get<JsonMap>(node->value))`) uses the private method of the class that inherits `SerializableItem`, and that's why I declare `friend class detail::JsonHelper;`

Comment (tigercosmos) [reply cpp/modmesh/serialization/SerializableItem.hpp]: > I don't like the output argument. But it looks like necessary evil here. Is it so? If yes, is it possible to make it private?

based on the current design, it is necessary.
What do you mean "private"? this is a helper class, and the function is static.
`SerializableItem::from_json` calls this function, and feeds the class member throw the output argument.

File: cpp/modmesh/serialization/SerializableItem.hpp
Comment (yungyuc) [reply cpp/modmesh/serialization/SerializableItem.hpp]: My question is against `else`.  Does it need `constexpr`?

File: cpp/modmesh/serialization/SerializableItem.hpp
Comment (tigercosmos) [reply cpp/modmesh/serialization/SerializableItem.hpp]: In addition, since it's a helper class (or basically a helper function), I think the length should be fine? we have much longer helper class in `array_common.hpp`

File: cpp/modmesh/serialization/SerializableItem.hpp
Comment (tigercosmos) [reply cpp/modmesh/serialization/SerializableItem.hpp:159]: therefore, I think it's necessary here.

File: cpp/modmesh/serialization/SerializableItem.hpp
Comment (yungyuc) [reply cpp/modmesh/serialization/SerializableItem.hpp]: > > And this function from_json_string is certainly too long for declaration. Please make it outside the class declaration.
> 
> I put it in the class because L228 (when the parser doing recursion) uses the private method of `SerializableItem`, and that's why I declare `friend class detail::JsonHelper;`

The access is OK.  Long functions are not easy to read when putting in class declaration.  They should be taken outside the declaration while still made inline.

File: cpp/modmesh/serialization/SerializableItem.hpp
Comment (tigercosmos) [reply cpp/modmesh/serialization/SerializableItem.hpp]: this function is okay to put outside, since it doesn't use the private function. However, as `static void from_json_string(const detail::JsonNodePtr & node, T & value)` must be put in the helper class, I think we should put this function in the helper class for the consistency?

File: cpp/modmesh/serialization/SerializableItem.hpp
Comment (yungyuc) [reply cpp/modmesh/serialization/SerializableItem.hpp]: > > I don't like the output argument. But it looks like necessary evil here. Is it so? If yes, is it possible to make it private?
> 
> based on the current design, it is necessary. What do you mean "private"? this is a helper class, and the function is static. `SerializableItem::from_json` calls this function, and feeds the class member throw the output argument.

It will be easier to maintain a function of output argument when it is private.  It does not matter the function to be static or not.  But if you use the class as a namespace for collecting static member functions and call them from other class, making them private is not realistic.

Please leave a code comment to warn maintainers of the output argument.  We should get rid of it in the future.

File: gtests/test_nopython_serializable.cpp
Comment (tigercosmos) [reply gtests/test_nopython_serializable.cpp:170]: sure, should be in the future PRs.

File: cpp/modmesh/serialization/SerializableItem.cpp
Comment (tigercosmos) [reply cpp/modmesh/serialization/SerializableItem.cpp:185]: fixed.

File: cpp/modmesh/serialization/SerializableItem.hpp
Comment (yungyuc) [reply cpp/modmesh/serialization/SerializableItem.hpp]: By putting outside the (class) declaration, I meant to move the function body, like:

```cpp
class C
{
    void F();
};
void C::F()
{
    // do something
}
```

It has nothing to do with function access.

File: cpp/modmesh/serialization/SerializableItem.cpp
Comment (tigercosmos) [reply cpp/modmesh/serialization/SerializableItem.cpp]: sorry, fixed.

Comment (tigercosmos) [reply cpp/modmesh/serialization/SerializableItem.cpp]: I hope clang-format can help for this.

File: cpp/modmesh/base.hpp
Comment (tigercosmos) [reply cpp/modmesh/base.hpp]: thanks for telling me

File: cpp/modmesh/base.hpp
Comment (tigercosmos) [reply cpp/modmesh/base.hpp:90]: note that `is_integral` doesn't not support `long long it`, and the support is until C++20 `std::integral`, but I think should be fine here.

File: cpp/modmesh/serialization/SerializableItem.hpp
Comment (tigercosmos) [reply cpp/modmesh/serialization/SerializableItem.hpp:237]: however,  the function is a template function, I don't think we can do it?

Comment (tigercosmos) [reply cpp/modmesh/serialization/SerializableItem.hpp]: I see, I realize it's possible to do it even for templates.

File: cpp/modmesh/serialization/SerializableItem.hpp
Comment (tigercosmos) [reply cpp/modmesh/serialization/SerializableItem.hpp]: > My question is against else. Does it need constexpr?

No need.

File: cpp/modmesh/serialization/SerializableItem.hpp
Comment (tigercosmos) [reply cpp/modmesh/serialization/SerializableItem.hpp]: fixed

**Comment by tigercosmos** [comment]
@yungyuc 
> Use STL isalpha, isdigit, is_integral, and is_floating_point.

all fixed.

>  Do not alias smart pointer types.

fixed.

> Use the modern using syntax.

fixed.

> Clarify how to improve the output argument of from_json_string().

output argument is needed by design.

> from_json_string() is too long for class declaration.

splitted.

All issues are addressed, please also check the comments.

File: cpp/modmesh/serialization/SerializableItem.cpp
Comment (yungyuc) [reply cpp/modmesh/serialization/SerializableItem.cpp:185]: This is good. Thanks.

File: cpp/modmesh/serialization/SerializableItem.cpp
Comment (yungyuc) [reply cpp/modmesh/serialization/SerializableItem.cpp]: If it does not, we can consider to craft a linter for ourselves.  In the future, of course.

File: cpp/modmesh/serialization/SerializableItem.hpp
Comment (yungyuc) [inline cpp/modmesh/serialization/SerializableItem.hpp:180]: Need code comment for the plan to mitigate the output arguments.

File: cpp/modmesh/serialization/SerializableItem.hpp
Comment (yungyuc) [inline cpp/modmesh/serialization/SerializableItem.hpp:252]: Ditto.  Output argument.

**Review by yungyuc — CHANGES_REQUESTED** [review]
- [x] Leave code comments for a plan to mitigate the output argument.

File: cpp/modmesh/serialization/SerializableItem.hpp
Comment (yungyuc) [reply cpp/modmesh/serialization/SerializableItem.hpp:180]: (I hit review changes button too quickly.)  We can discuss here or discord for how to author the plan.  It may be a simple one.

File: cpp/modmesh/serialization/SerializableItem.hpp
Comment (tigercosmos) [reply cpp/modmesh/serialization/SerializableItem.hpp]: @yungyuc  based on the current design:
in `from_json_string`:
```cpp
    else if constexpr (std::is_base_of_v<SerializableItem, T>)
    {
        if (node->type != detail::JsonType::Object)
        {
            throw std::runtime_error("Invalid JSON format: invalid object type.");
        }
        // NOLINTNEXTLINE(misc-no-recursion)
        value.from_json(std::get<JsonMap>(node->value)); /* recursive here */
    }
 ```
 
Note that we call `value`'s mutable method `from_json`. Therefore, we must pass the **mutable** reference of `value` as the argument of `from_json_string`.
 
 I don't think it can be an easy change to get rid of the output argument unless we change the whole design.
 
 Also,  I didn't fully understand why you wanted to remove the output argument. This function is live in the `detail` namespace, and the implementation is hidden under the macro, I don't think the user has chance to use this function.

**Comment by tigercosmos** [comment]
@yungyuc  please see my comment here: https://github.com/solvcon/modmesh/pull/377#discussion_r1653774032

File: cpp/modmesh/serialization/SerializableItem.hpp
Comment (yungyuc) [reply cpp/modmesh/serialization/SerializableItem.hpp]: It's a good point that the static member function is in `detail`.  I missed it.  In this case I agree the output argument is not so severe.  Could you please add a simple code comment indicating the output arguments are bad for maintenance and should be fixed in the future?  (A concrete plan does not need to be mentioned.)

It's quite possible for a maintainer to move a useful helper class outside `detail`, in the future.  Then it will become an issue.

**Comment by tigercosmos** [comment]
@yungyuc  comments for the output argument is added.

File: cpp/modmesh/serialization/SerializableItem.hpp
Comment (yungyuc) [inline cpp/modmesh/serialization/SerializableItem.hpp]: Thanks, but please beautify the comments.  Appending after the signature makes the lines too line.  Please add the comments before the signature in distinct lines.

**Review by yungyuc — CHANGES_REQUESTED** [review]
Thanks, @tigercosmos .  We need some cosmetics.

- [x] Cosmetic changes for the comments.

**Comment by yungyuc** [comment]
I saw the change and will merge once CI completes.

**Review by yungyuc — APPROVED** [review]

**Pull request merged**
